File: 12/main.js
#!/usr/bin/env node
const DIRECTIONS = [
  [ 1,  0], // East.
  [ 0,  1], // North.
  [-1,  0], // West.
  [ 0, -1], // South.
];
DIRECTIONS.E = DIRECTIONS[0];
DIRECTIONS.N = DIRECTIONS[1];
DIRECTIONS.W = DIRECTIONS[2];
DIRECTIONS.S = DIRECTIONS[3];


const mod4 = v => (4 + (v % 4)) % 4;


const ACTIONS_ONE = {
  N(ship, amount) {
    ship.y += amount;
  },
  S(ship, amount) {
    this.N(ship, -amount);
  },
  E(ship, amount) {
    ship.x += amount;
  },
  W(ship, amount) {
    this.E(ship, -amount);
  },
  L(ship, degrees) {
    ship.direction += (degrees / 90);
  },
  R(ship, degrees) {
    this.L(ship, -degrees);
  },
  F(ship, amount, direction) {
    const [dx, dy] = DIRECTIONS[mod4(ship.direction)];
    ship.x += amount * dx;
    ship.y += amount * dy;
  },
};


const ROTATIONS = [
  [ 1,  0,
    0,  1],
  [ 0, -1,
    1,  0],
  [-1,  0,
    0, -1],
  [ 0,  1,
   -1,  0],
].map(rot => new Int32Array(rot));


const ACTIONS_TWO = {
  N(ship, amount) { ACTIONS_ONE.N(ship.wp, amount); },
  S(ship, amount) { ACTIONS_ONE.S(ship.wp, amount); },
  E(ship, amount) { ACTIONS_ONE.E(ship.wp, amount); },
  W(ship, amount) { ACTIONS_ONE.W(ship.wp, amount); },
  L(ship, degrees) {
    const rot = ROTATIONS[mod4(degrees / 90)];
    const x = ship.wp.x;
    const y = ship.wp.y;
    [ship.wp.x, ship.wp.y] = [
      rot[0] * x + rot[1] * y,
      rot[2] * x + rot[3] * y,
    ];
  },
  R(ship, amount) {
    this.L(ship, -amount);
  },
  F(ship, amount) {
    ship.x += amount * ship.wp.x;
    ship.y += amount * ship.wp.y;
  },
};


const ships = require('fs')
  .readFileSync(
    process.argv[2] || require('path').join(__dirname, 'input.txt'),
    'utf8'
  )
  .trim()
  .split('\n')
  .map(s => ({
    action: s[0],
    amount: +s.substr(1),
  }))
  .reduce((ships, {action, amount}) => {
    ACTIONS_ONE[action](ships[0], amount);
    ACTIONS_TWO[action](ships[1], amount);
    // console.debug(action, amount, '\t', ships[1]);
    return ships;
  }, [
    {x: 0, y: 0, direction: 0, wp: undefined},
    {
      x: 0,
      y: 0,
      wp: {x: 10, y: 1, direction: 0},
    },
  ]);
// console.debug(ships);
console.log('Part one:', Math.abs(ships[0].x) + Math.abs(ships[0].y));
console.log('Part two:', Math.abs(ships[1].x) + Math.abs(ships[1].y));

File: 12/main.cpp
#include <cmath>
#include <iostream>
#include <fstream>


struct Matrix {
  int a; int b;
  int c; int d;
};


Matrix rotations[4] = {
  { 1,  0,
    0,  1},
  { 0, -1,
    1,  0},
  {-1,  0,
    0, -1},
  { 0,  1,
   -1,  0},
};


struct Point {
  int x;
  int y;

  Point& operator+=(Point a) noexcept {
    x += a.x;
    y += a.y;
    return *this;
  }
};


Point operator*(Matrix const& m, Point const p) noexcept {
  return {m.a * p.x + m.b * p.y, m.c * p.x + m.d * p.y};
}

Point operator*(int scalar, Point const p) noexcept {
  return {scalar * p.x, scalar * p.y};
}


int taxicab(Point const p) noexcept {
  return std::abs(p.x) + std::abs(p.y);
}


int mod4(int v) {
  return (4 + (v % 4)) % 4;
}


struct Rotateable {
  Point location;
  int direction;
};


struct Ship {
  Rotateable one;
  struct PartTwo {
    Rotateable waypoint;
    Point location;
  } two;
};


void north(Ship& ship, int amount) {
  ship.one.location.y += amount;
  ship.two.waypoint.location.y += amount;
}

void east(Ship& ship, int amount) {
  ship.one.location.x += amount;
  ship.two.waypoint.location.x += amount;
}

void south(Ship& ship, int amount) {
  ship.one.location.y -= amount;
  ship.two.waypoint.location.y -= amount;
}

void west(Ship& ship, int amount) {
  ship.one.location.x -= amount;
  ship.two.waypoint.location.x -= amount;
}

void left(Ship& ship, int degrees) {
  int const turns = degrees / 90;
  ship.one.direction += turns;
  auto const& rot = rotations[mod4(turns)];
  ship.two.waypoint.location = rot * ship.two.waypoint.location;
}

void right(Ship& ship, int degrees) {
  left(ship, -degrees);
}

void forward(Ship& ship, int amount) {
  auto const& rot = rotations[mod4(ship.one.direction)];
  ship.one.location += rot * Point{amount, 0};
  ship.two.location += amount * ship.two.waypoint.location;
}


#ifndef NDEBUG
template <typename OS>
OS& operator<<(OS& os, Point p) {
  return os << "{x: " << p.x << ", " << "y: " << p.y << '}';
}


template <typename OS>
OS& operator<<(OS& os, Rotateable const& r) {
  return os << '{' << r.location << ", d: " << r.direction << '}';
}
#endif


int main(int const argc, char const* const* const argv) {
  std::ifstream ifs(argc < 2 ? DEFAULT_INPUT_FILE : argv[1]);

  Ship ship{
    Rotateable{
      Point{0, 0},
      0},
    Ship::PartTwo{
      Rotateable{
        Point{10, 1},
        0},
      Point{0, 0}}};

  for (char action = ifs.get(); ifs.good(); action = ifs.get()) {
    int amount;
    ifs >> amount >> std::ws;
    // Not sure it's worth pre-parsing the input today.
    switch (action) {
      case 'N': north(ship, amount); break;
      case 'S': south(ship, amount); break;
      case 'E': east(ship, amount); break;
      case 'W': west(ship, amount); break;
      case 'L': left(ship, amount); break;
      case 'R': right(ship, amount); break;
      case 'F': forward(ship, amount); break;
      default:
        std::cerr << "Invalid action: " << action << std::endl;
        return 1;
    }

#ifndef NDEBUG
    std::cout << action << ' ' << amount << '\t'
      << ship.two.location << '\t'
      << ship.two.waypoint << std::endl;
#endif
  }

  std::cout << "Part one: " << taxicab(ship.one.location) << '\n'
            << "Part two: " << taxicab(ship.two.location) << std::endl;

  return 0;
}
